mono/metadata/handle.c

```diff
@@ -0,0 +1,252 @@
+ */
+MonoHandle*
+mono_handle_arena_elevate (MonoHandleArena *arena, MonoHandle *handle)
+{
+	g_assert (handle);
+	g_assert (arena);
+	g_assert (arena->prev);
+
+	return handle_new (arena->prev, handle->obj);
+}
+
+MonoHandle*
+mono_handle_elevate (MonoHandle *handle)
+{
+	MonoInternalThread *thread;
+	MonoHandleArena *arena;
+
+	thread = mono_thread_internal_current ();
+	g_assert (thread);
+
+	arena = thread->handle_arena;
+
+	return mono_handle_arena_elevate (arena, handle);
+}
+
+gsize
+mono_handle_arena_size (gsize nb_handles)
+{
+	g_assert (nb_handles > 0);
+	return sizeof (MonoHandleArena) + sizeof (MonoHandleArenaChunk) + sizeof (MonoHandle) * (MAX (nb_handles, HANDLES_PER_CHUNK) - MONO_ZERO_LEN_ARRAY);
+}
+
+void
+mono_handle_arena_stack_push(MonoHandleArena **arena_stack, MonoHandleArena *arena, gsize nb_handles)
+{
+	g_assert (arena_stack);
+	g_assert (arena);
+
+	arena->prev = *arena_stack;
+	arena->chunk = arena->chunk_last = (MonoHandleArenaChunk*) (((char*) arena) + sizeof (MonoHandleArena));
+
+	arena->chunk->next = NULL;
+	arena->chunk->handles_capacity = MAX (nb_handles, HANDLES_PER_CHUNK);
+	arena->chunk->handles_size = 0;
+	memset (&arena->chunk->handles [0], 0, sizeof (MonoHandle) * arena->chunk->handles_capacity);
+
+	*arena_stack = arena;
+}
+
+void
+mono_handle_arena_push (MonoHandleArena *arena, gsize nb_handles)
+{
+	MonoInternalThread *thread;
+
+	thread = mono_thread_internal_current ();
+	g_assert (thread);
+
+	mono_handle_arena_stack_push (&thread->handle_arena, arena, nb_handles);
+}
+
+void
+mono_handle_arena_stack_pop(MonoHandleArena **arena_stack, MonoHandleArena *arena, gsize nb_handles)
+{
+	MonoHandleArenaChunk *chunk, *next;
+
+	g_assert (arena);
+	g_assert (arena->chunk);
+	g_assert (arena->chunk_last);
+	g_assert (arena_stack);
+
+	*arena_stack = arena->prev;
+
+	for (chunk = arena->chunk; chunk; chunk = next) {
+		next = chunk->next;
+		memset (&chunk->handles [0], 0, sizeof (MonoHandle) * chunk->handles_capacity);
+		if (chunk != arena->chunk)
+			chunk_free (chunk);
+	}
+}
+
+void
+mono_handle_arena_pop (MonoHandleArena *arena, gsize nb_handles)
+{
+	MonoInternalThread *thread;
+	thread = mono_thread_internal_current ();
+	g_assert (thread);
+
+	mono_handle_arena_stack_pop (&thread->handle_arena, arena, nb_handles);
+}
+
+static void
+arena_scan (gpointer addr, MonoGCMarkFunc mark_func, gpointer gc_data)
+{
+	MonoHandleArena *arena;
+	MonoHandleArenaChunk *chunk;
+	int i;
+
+	for (arena = *(MonoHandleArena**) addr; arena; arena = arena->prev) {
+		for (chunk = arena->chunk; chunk; chunk = chunk->next) {
+			for (i = 0; i < chunk->handles_size; ++i) {
+				if (chunk->handles [i].obj != NULL)
+					mark_func (&chunk->handles [i].obj, gc_data);
+			}
+		}
+	}
+}
+
+static void
+initialize (void)
+{
+#ifdef HAVE_SGEN_GC
+	arena_desc = mono_gc_make_root_descr_user (arena_scan);
+#endif
+}
+
+void
+mono_handle_arena_init_thread (MonoInternalThread *thread)
+{
+#ifdef HAVE_SGEN_GC
```

Comment:
I don't follow.  Every `MonoHandleArenaChunk` is allocated using malloc, and `MonoHandle*` is just an interior pointer into a chunk.  Any live `MonoHandle*` in mutator code will be reachable by Boehm from the stack.  I don't think there's anything to register.